xdsl/backend/riscv/register_rvscf_regalloc.py

```diff
@@ -0,0 +1,201 @@
+        if self.liveliness[val.type] > 0:
+            # this may be interesting behaviour? idk
+            pass
+        self.liveliness[val.type] = len(val.uses)
+
+    def register_use(self, val: SSAValue):
+        assert isinstance(val.type, riscv.RISCVRegisterType) and val.type.is_allocated
+        if self.liveliness[val.type] > 0:
+            self.liveliness[val.type] -= 1
+
+    def make_forbidden(self, reg: riscv.RISCVRegisterType):
+        self.forbidden_vals.add(reg)
+
+
+def register_allocate_function(func: riscv_func.FuncOp):
+    # create an empty context
+    ctx = RegAllocCtx()
+
+    # register all function args and their liveliness
+    for arg in func.body.blocks[0].args:
+        assert isinstance(arg.type, riscv.RISCVRegisterType)
+        assert arg.type.is_allocated
+        ctx.add_reg(arg)
+
+    # allocate the body
+    register_allocate_region(func.body, ctx)
+
+
+def register_allocate_for_op(op: riscv_scf.ForOp, ctx: RegAllocCtx):
+    """
+    Register allocate the for loop
+
+    Take care of:
+        - loop counter stays in same register
+        - loop carried variables stay in same register
+    """
+    # construct an inner contex for the loop body
+    inner_ctx = ctx.child_for_loops()
+
+    # make sure all loop carried variables stay in the same registers during the loop:
+    # gently force the block argument and yielded values in the same register as the
+    # given values
+    yield_op = op.body.block.last_op
+    assert isinstance(yield_op, riscv_scf.YieldOp)
+    for input_val, block_arg, yielded, ret_val in zip(
+        op.iter_args, op.body.block.args[1:], yield_op.operands, op.results
+    ):
+        block_arg.type = input_val.type
+        yielded.type = input_val.type
+        ret_val.type = input_val.type
+
+    # if the lb has no other uses, use it as the loop counter otherwise
+    # grab a free register for the loop counter (from the inner context)
+    loop_counter_reg = op.lb.type if len(op.lb.uses) == 1 else inner_ctx.free_int_reg()
+    assert isinstance(loop_counter_reg, riscv.IntRegisterType)
+
+    iter_val = op.body.block.args[0]
+    iter_val.type = loop_counter_reg
+    inner_ctx.register_use(iter_val)
+    # make the iter_val forbidden so it is not overwritten in the loop
+    inner_ctx.make_forbidden(loop_counter_reg)
+
+    register_allocate_region(op.body, inner_ctx)
+
+
+def register_allocate_region(reg: Region, ctx: RegAllocCtx):
+    """
+    Iterate over all ops in a region and register allocate them
+    """
+    for block in reg.blocks:
+        for op in block.ops:
+            if isinstance(
+                op,
+                riscv.RISCVOp,
+            ):
+                for result in op.results:
+                    assert isinstance(result.type, riscv.RISCVRegisterType)
+                    # keep track of the "alive" registers
+                    for operand in op.operands:
+                        ctx.register_use(operand)
```

Comment:
Is there a reason why this register allocator uses a different strategy for Blocks than `RegisterAllocatorLivenessBlockNaive`? I had hoped that our scf allocator would just be an extended version of `RegisterAllocatorLivenessBlockNaive`. Is there a reason we choose a different approach here?